### DISCUSSION TITLE
Forward over reverse off-diagonal Hessian-vector product

### DISCUSSION THREAD

**Discussion by qsh-zh** [jax-ml/jax#9294] [Q&A]
I have a scalar function 
![image](https://user-images.githubusercontent.com/28699575/150704940-a4275ef5-9114-43ea-b3d1-5c53b119cd6d.png)

 I am interested in the following term
![image](https://user-images.githubusercontent.com/28699575/150705518-b0b498f4-8b65-4e27-912d-eca5e334113a.png)

My questions as follows:

1. "forward-over-reverse" off-diagonal Hessian-vector product. According to the [tutorial](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#hessian-vector-products-using-both-forward-and-reverse-mode), we should have an efficient approach based on forward-over-reverse, but how to do it for this off -digonal Hessian matrix, 
2.  The vector part and hessian are highly corrected, can we save some calculation? For example, save get gradients when calculating the hessian term?

**Comment by YouJiacheng** [comment]
1. See below.
2. Yes, jvp over grad just do this. But for hessian-grad product, we need to use linearize.
```python
def f(y, x):
    return jnp.sum(y) + jnp.sum(x) # just example

def g(y, x, u):
    f_x_func = jax.grad(f, argnums=1) # autodiff for x only
    f_x, f_yx_dot_u = jax.jvp(lambda _y: f_x_func(_y, x), (y,), u)  # use closure to autodiff for y only
    # \frac{\partial f(y,x)}}{\partial x} and \frac{\partial^2 f(y,x)}}{\partial y\partial x} u

def h(y, x, u):
    f_x_func = jax.grad(f, argnums=1) # autodiff for x only
    f_x, f_yx_func = jax.linearize(lambda _y: f_x_func(_y, x), y)  # use closure to autodiff for y only
    return f_yx_func(f_x - u)
```

**Reply by qsh-zh** [reply]
@YouJiacheng Thanks for the clear explanation.

**Reply by qsh-zh** [reply]
@YouJiacheng Sorry, I think there is a small mistake in your answer. In the last line, `f_yx_func` can work on primal value, whose shape should similar to the y.shape.

**Reply by YouJiacheng** [reply]
> @YouJiacheng Sorry, I think there is a small mistake in your answer. In the last line, `f_yx_func` can on work on primal value, whos shape should similar to the y.shape.

Oh, I just rephrase your formula and not see that \frac{\partial^2 f(y,x)}}{\partial y\partial x} has shape m×n.
I think if \frac{\partial^2 f(y,x)}}{\partial y\partial x} is \frac{\partial}{\partial y}\frac{\partial f(y,x)}}{\partial x} it can not dot with tangent with shape==x.shape.
I think you can change the order of differentiation, or use
```python
def h(y, x, u):
    f_x_func = jax.grad(f, argnums=1) # autodiff for x only
    f_x, f_yx_func = jax.linearize(lambda _y: f_x_func(_y, x), y)  # use closure to autodiff for y only
    return jax.linear_transpose(f_yx_func, y)(f_x - u) # linear_transpose only use y.shape and y.dtype
```

**Reply by qsh-zh** [reply]
But actually, if we use `linear_transpose`, it is equivalent to reverse over the reverse. Seems forward over reverse is not a good option here.

**Reply by YouJiacheng** [reply]
> But actually, if we use `linear_transpose`, it is equivalent to reverse over the reverse. Seems forward over reverse is not a good option here.

I agree that `linearize` + `linear_transpose` is equivalent to `vjp`(but `linear_transpose` itself isn't). Actually, `linearize` costs about same memory as reverse-mode AD.

I think it is efficient to change the partial differentiation order(assume C^2 smooth), and sequentially evaluate the gradient value using a separate call to grad function and the hessian-vector product value using `jvp`(abort the out primal). The redundancy is caused by  the sequential dependence between hessian-vector product and evaluate gradient since your vector contain the gradient.
